### ISSUE TITLE
545407F2 - Grand Theft Auto IV: Complete Edition

### ISSUE THREAD

**Issue by mavethee** [xenia-canary/game-compatibility#14]
[Marketplace](https://marketplace.xbox.com/en-us/Product/GTA-IV/66acd000-77fe-1000-9115-d802545407f2)

Tested on https://github.com/xenia-canary/xenia-canary/commit/5d39e520e32a42af18e56b8b6af642a70f3a4716

# Issues:
GTA IV (Disc 1) has still gameplay state, game saves but can't load save. Works with TU and gives a few more fps with it. Episodes from Liberty City (Disc 2) doesn't detect properly TU even if MediaID matches. Without TU, goes ingame, at least TBoGT because TLaD fails at loading screen.
# Log:
**_Disc 1:_**
[xenia-GTAIVCEDisc1withTU.zip](https://github.com/xenia-canary/game-compatibility/files/4025555/xenia-GTAIVCEDisc1withTU.zip)
**_Disc 2:_**
[xenia-GTAIVCEDisc2withoutTU-TBoGT.zip](https://github.com/xenia-canary/game-compatibility/files/4025562/xenia-GTAIVCEDisc2withoutTU-TBoGT.zip)
[xenia-GTAIVCEDisc2withoutTU-TLaD.zip](https://github.com/xenia-canary/game-compatibility/files/4025589/xenia-GTAIVCEDisc2withoutTU-TLaD.zip)

# Screenshot(s):
![unknown](https://user-images.githubusercontent.com/47083527/71810349-7c6acb00-3072-11ea-883e-9a50d91a640a.png)
![unknown (1)](https://user-images.githubusercontent.com/47083527/71810354-7f65bb80-3072-11ea-8e97-3f4359ff5ff1.png)
![unknown (2)](https://user-images.githubusercontent.com/47083527/71810358-81c81580-3072-11ea-953b-d9d6bc2ebb02.png)
![unknown (3)](https://user-images.githubusercontent.com/47083527/71810361-84c30600-3072-11ea-80ec-a3e1bb6413b2.png)
![unknown (4)](https://user-images.githubusercontent.com/47083527/71810420-a7edb580-3072-11ea-8e46-81bd2a4e8b5d.png)
![unknown (5)](https://user-images.githubusercontent.com/47083527/71810429-ab813c80-3072-11ea-9ac1-6c098fb4b919.png)
![unknown (6)](https://user-images.githubusercontent.com/47083527/71811412-e8e6c980-3074-11ea-95c8-f79480eace0d.png)

# Labels:
state-gameplay, tech-engine-rage.

**Comment by emoose** [comment]
Hm, AFAIK the TU only applies to Disk 1, at least there's only the one xexp inside it...

Also with the latest canary build you should be able to keep the TU in the content folder for Disk 1 to use, and also be able to run Disk 2 without it failing to launch (patch will fail since it's not meant for Disk 2, but it should still launch fine)

**Comment by mavethee** [comment]
@emoose the thing is game uses different MediaIDs for both discs. The is not like in RDR case when GOTY Edition contains update for both discs. Plus something borked with The Lost and Damned, you able to look at it? 😅

**Comment by emoose** [comment]
I guess either Disk 2 is already patched with latest TU (the ISO seems to contain an update folder at least), or maybe it uses a different TU file to Disk 1 or something like that..

I'll try taking a look at TLAD soon, dunno if I'll be able to figure anything out though.

**Comment by CrazyCSIW6** [comment]
2 years later, saving problems are fixed on latest Canary builds, only issues now are with crashes which are very frequent in this game for some reason.
I noticed in TLAD the game would crash basically 9 times out of 10 on the opening cutscenes and I had to skip them to progress, don't know if any other cutscenes crash in the same way though.
I'd say this game is currently playable with the exception of crashes

**Comment by betelopson** [comment]
I assume if your game crashes then you have AMD hardware like CPU and GPU so change this in "xenia-canary.config.toml":

gpu = "vulkan"

This will fix the crashes completely but no upscale option with Vulkan.

This way the game is perfectly works. I've played only the single disc version of the game, but everything worked perfectly until a mission where we need to take care of a situation, and we need to choose who will be shot and why is that because transparency does not work on grass, fences, far away bridges and other weird full-screen effects.

**Comment by A1eNaz** [comment]
## Status
`state-playable`

### Screenshot(s)
![Screenshot_6](https://github.com/user-attachments/assets/a094fb39-2a3a-49e5-ba6e-e4a813a07fdb)

### Config changes
```
gpu = "vulkan"
(Prevents the game from freezing but vegetation becomes blocky.)

mount_cache = true
(Adds missing objects in the map.)
```

## Issues
- Not able to get in the helicopter during the `The Final Commission` which can be fixed by capping the game to 30 FPS. **(Not an emulator bug.)**
- Some buildings or subways are pitch black. **(Most notably when you go play pool.)**

### Notes
- No game-breaking issues were encountered during my playthrough except the `The Final Commission` bug.

**Comment by Sowa95** [comment]
https://github.com/xenia-canary/xenia-canary/commit/60318a5db66516614295146e5d455e4b6fda1be7

[xenia.zip](https://github.com/user-attachments/files/19125160/xenia.zip)

apu-garbage | requires_mount_cache | regression

### Mount_cache

If mount_cache enabled, GTA IV writes a bunch stuff to `cache` and `cache1`. Surprisingly, the files are created live in gameplay. After driving around the city for about 15 minutes, combined size of caches capped at ~3GB. If X360 version shares any similarities with PS3 version (mandatory installation on PS3), caches could grow up to ~3.3GB.

I suspect GTA IV uses different streaming approach if HDD cache is available, allowing for more objects around the world to be rendered. This is not draw distance, additional objects will never render if mount_cache isn't enabled.

<details><summary>mount_cache = true vs false</summary>
<p>

![Image](https://github.com/user-attachments/assets/a79c4fa0-6443-4418-bc5e-eedfdb1f0f38)
![Image](https://github.com/user-attachments/assets/39995ffa-af86-46e9-ba15-9c62c54874a2)

</p>
</details>

### 16580f5 audio regression

There's a few audio issues caused by 16580f5 which carry over to `old` and `new` XMA decoders. Busted car radio is the most noticeable issue. Some tracks end too soon, and there's an echo/muffle that goes on and off, perhaps something wrong with channels. `Master` XMA decoder doesn't have this issue, but I didn't test it thoroughly, there could be some other issues elsewhere.

<details><summary>Details</summary>
<p>

Music in intro (Rockstar logos):
c9cd6f1 - music is fine.
from 16580f5 to 97fdf9c - music is very muffled. It cuts out before loading screen runs out.
b2dd489 - music is slightly muffled, doesn't cut off. [YT](https://youtu.be/fav3_ddQbfU)

Echoed/muffled radio:
c9cd6f1 - radio seems fine, I haven't noticed anything in 30 minutes. [YT](https://youtu.be/nB5mSut_dik)
from 16580f5 to 97fdf9c - radio is a complete mess. Very muffled, tracks cut off way too early. [YT](https://youtu.be/CKW0WHeD0Zw)
b2dd489 - muffle/echo in the radio still present, but not as strong. Tracks can cut off, but rarely. [YT](https://youtu.be/Mk19ke_S70A)

Car horn doesn't loop when holding LS. This is an old issue, but the pause between honks got longer in some cars, like Ranger:
97fdf9c - no pause between horn, almost continuos.
from b2dd489 to 668244a - very long pause between horn.
d24d329 - short pause between horn. [YT](https://youtu.be/GaWTC6mGy5k)

Save with parked Ranger:
[545407F2.zip](https://github.com/user-attachments/files/19125282/545407F2.zip)

</p>
</details> 

### Nvidia driver regression (outdated)

Driver 526.47 broke lighting in some locations. It has been fixed in driver 595.71

<details><summary>Details</summary>
<p>

522.25 - last driver that looks fine
526.47 to 591.86 - some areas are too dark, Vulkan-FSI helps.
595.71 - fixed

**522.25 vs 526.47**
![Image](https://github.com/user-attachments/assets/02d57e21-1e96-4ab7-bb45-7ec65ae6d436)

**591.86 vs 595.71**
<img width="3841" height="1078" alt="Image" src="https://github.com/user-attachments/assets/334963fe-c354-4c24-9e92-486d6daf1f6c" />

</p>
</details>

**Comment by Spadium** [comment]
The game on Linux is not stable, it crashes with disc read errors, and it doesn't crash at the same place until I get to the safehouse where it consistently crashes on loading the save game. There is also no audio after the load screens before starting a new game on both xma decoders, with some popping in gameplay, cutscenes, and exiting the emulator. I'll try copying the game from my Xbox again and see if that changes anything, but these issues don't show up on my Windows 11 install, which plays the game with no noticeable issues from the same USB drive as Xenia on my Linux install.

![Image](https://github.com/user-attachments/assets/0b70b4c0-d2aa-4887-8831-d2e25c57b8f5)

**Comment by RiverDev12** [comment]
mine had none but i do have a PC windows RTX 5050 with 65gb ram and 750w power supply running off SSD so very quick 

I only had errors on my gammig laptop so grab a pc for this one

**Comment by nfxw** [comment]
Dark interior is now fixed on latest canary build.

<img width="1920" height="1200" alt="Image" src="https://github.com/user-attachments/assets/6ea30d4a-ce94-4ad9-a384-7302b72db38e" />

**Comment by NWBArk** [comment]
Driver: 591.74
GPU: RTX 4090

Tested on https://github.com/xenia-canary/xenia-canary/commit/263e91b014ed3dc4e75b4f1fe23bd0388f23fea2 with d3d12, the interior lighting is still broken:

<img width="1282" height="772" alt="Image" src="https://github.com/user-attachments/assets/7476fa97-0d77-4f2d-8c8e-ffe19e334d87" />
<img width="1282" height="772" alt="Image" src="https://github.com/user-attachments/assets/bb5357f3-0754-4ea4-9b9d-cdb19d4f8e7d" />

**Comment by NWBArk** [comment]
> Might be fixed on Edge with Vulkan, this retard has a habit of not providing any useful information. No log, no title bar, go figure. He might as well be on Vulkan-FSI.

This interior lighting issue can't be observed on either latest canary nor edge build with Vulkan. But I do believe edge fixed the black transparency issue with trees/fences.

Also sorry for forgetting to attach the log, I'll upload it to the original comment once I have access to my PC.

**Comment by NWBArk** [comment]
> [@NWBArk](https://github.com/NWBArk) can you upload screenshot from main Canary? I just tried Vulkan-FBO and it's still dark in the pool bar.

Both Canary and Edge builds have this issue with Vulkan FBO but not with FSI:

https://github.com/xenia-canary/xenia-canary/commit/263e91b014ed3dc4e75b4f1fe23bd0388f23fea2
<details>
<summary>Vulkan FBO & FSI</summary>
FBO:
<img width="1282" height="772" alt="Image" src="https://github.com/user-attachments/assets/4ea98d70-615a-45e5-8f6e-200175c927b7" />

FSI:
<img width="1282" height="772" alt="Image" src="https://github.com/user-attachments/assets/f48388ac-e7f7-4173-8085-af9e9d8b6ded" />
</details>

https://github.com/has207/xenia-edge/commit/f77c654e77496d697da3521485b6ed11deb08cb3
<details>
<summary>Vulkan FBO & FSI</summary>
FBO:
<img width="1282" height="752" alt="Image" src="https://github.com/user-attachments/assets/2942d6bc-2144-4dbe-8442-bab9dc7cc64e" />

FSI:
<img width="1282" height="752" alt="Image" src="https://github.com/user-attachments/assets/dda206b6-0c3f-4cb7-82b9-c062b1a41e9d" />

</details>

The Edge build I'm using is not the latest because in https://github.com/has207/xenia-edge/commit/77bb2ea2d9ed2f4ec522b8e1e84ffa6d151039ab some performance regression was introduced to Vulkan.

Also I couldn't help but notice only <b>shadow-casting local lights</b> have this issue. Might be something to do with how D3D12/Vulkan FBO treats depth stencil buffer?

With D3D12 RTV this issue is somewhat leviated albeit not entirely.

**Comment by nfxw** [comment]
@Sowa95 
First thing first: if you called someone on this or other thread "retard" or "clown", then your message will be marked as spam.
Second: I've always use directx 12(which is shown in osd), RTV.
Third: I've tested episodes, because in fight club all interiors were black, now this issue is fixed, they both share the same "rage" engine, so i send this to both threads.

**Comment by matiashuarta** [comment]
Game at 2x resolution bugs map and left side of the screen with vertical line but renders darker than it should be.

Game at 1x resolution looks like crap natively and dark.
